Fog move to pixel shader

medievil1 opened the pull request.

medievil1 committed:
test fix for shader

jack said it should be fine to set it to program since passthrough shader is a program and get uploaded anyway

g

medievil1 committed:
fog stuff

medievil1 committed:
add FF pixelshader fog

jackchentwkh commented on src/core/hle/D3D8/Direct3D9/CxbxPixelShaderTemplate.hlsl:
```diff
@@ -334,6 +340,31 @@ float3 DoBumpEnv(const float4 TexCoord, const float4 BumpEnvMat, const float4 sr
+       fogFactor = (fogEnd - fogDepth) / (fogEnd - fogStart);
+	if (FOGENABLE == 0)
+	   fogFactor = 1;
+	   
```
I am not sure how flow control in HLSL should be. could this be revised to    
    if (FOGENABLE == 0)
    {
	   fogFactor = 1;
    }
    else
    {
       if(fogTableMode == FOG_TABLE_NONE) 
          fogFactor = fogDepth;
       if(fogTableMode == FOG_TABLE_EXP) 
          fogFactor = 1 / exp(fogDepth * fogDensity); // 1 / e^(d * density)
       if(fogTableMode == FOG_TABLE_EXP2) 
          fogFactor = 1 / exp(pow(fogDepth * fogDensity, 2)); // 1 / e^((d * density)^2)
       if(fogTableMode == FOG_TABLE_LINEAR) 
          fogFactor = (fogEnd - fogDepth) / (fogEnd - fogStart);
    }

jackchentwkh replied:
@medievil1 I am asking about flow control, not the formatting. did you try the flow control I listed above?

medievil1 replied:
changed and committed, though did not see any noticeable differences

jackchentwkh commented on src/devices/video/EmuNV2A_PGRAPH.cpp:
```diff
@@ -4815,8 +4815,8 @@ int pgraph_handle_method(
                         if (method_count==2){
```
the if (method_count==2){ in line 4815 could be removed as well, remember to remove the other "}" associated with this if statement as well.
there are only two vertex shader modes in PB, program and fixed mode. 
I would suggest to removed all unused code in this section and keep these two lines 
 g_NV2A_VertexShaderMode = VertexShaderMode::ShaderProgram;
g_NV2AVertexShader_dirty = true;

medievil1 committed:
fix formatting issues

address review comment

medievil1 committed:
fix formatting issues

address review comment

medievil1 committed:
remove no longer needed method count check

in program/FF vertex shader select
and remove old commented out code from same block.
Per Jacks suggestion

medievil1 committed:
crash fix....

everything in this branch crashes on startup unless there is some kind of delay here
Sleep(0) also works
key thing is the routine seems to try to run before everything is set

medievil1 committed:
crash fix....

everything in this branch crashes on startup unless there is some kind of delay here
Sleep(0) also works
key thing is the routine seems to try to run before everything is set

jackchentwkh commented on src/devices/video/EmuNV2A_PGRAPH.cpp:
```diff
@@ -4804,66 +4804,9 @@ int pgraph_handle_method(
-								 NV097_SET_TRANSFORM_EXECUTION_MODE_MODE_FIXED|NV097_SET_TRANSFORM_EXECUTION_MODE_RANGE_MODE_PRIV, method count 1)
-						to start fix function vertex shader program
-						*/
 					}else if (slot== NV097_SET_TRANSFORM_EXECUTION_MODE_MODE_FIXED){//fix function mode
```
please change line 4810 from 
}else if (slot== NV097_SET_TRANSFORM_EXECUTION_MODE_MODE_FIXED){//fix function mode
to 
}else{
  we don't need the test for  if (slot== NV097_SET_TRANSFORM_EXECUTION_MODE_MODE_FIXED) here any more.

medievil1 replied:
done....

medievil1 committed:
just adding more info to comment

medievil1 committed:
just adding more info to comment

medievil1 committed:
address review comment

medievil1 committed:
address review comment

medievil1 committed:
address review comment

changed fog enable/ fog table code flow

medievil1 committed:
address review comment

changed fog enable/ fog table code flow

medievil1 committed:
Merge branch 'PB-PR' of https://github.com/medievil1/Cxbx-Reloaded into PB-PR

medievil1 committed:
revert test fix for shader

removing for now
